meshroom/ui/components/thumbnail.py

```diff
@@ -0,0 +1,210 @@
+        When this is done, the createdThumbnail signal is emitted.
+
+        Args:
+            imgSource (QUrl): location of the input image
+
+        Returns:
+            QUrl: location of the corresponding thumbnail if it exists, otherwise None
+        """
+        if not imgSource.isValid():
+            return None
+
+        imgPath = imgSource.toLocalFile()
+        path = ThumbnailCache.thumbnailPath(imgPath)
+        source = QUrl.fromLocalFile(path)
+
+        # Check if thumbnail already exists
+        if os.path.exists(path):
+            # Update last modification time
+            Path(path).touch(exist_ok=True)
+            return source
+
+        # Thumbnail does not exist
+        # create it in a worker thread to avoid UI freeze
+        ThumbnailCache.pool.apply_async(self.createThumbnail, args=(imgSource,))
+        return None
+
+    def createThumbnail(self, imgSource):
+        """Load an image, resize it to thumbnail dimensions and save the result in the cache directory.
+
+        Args:
+            imgSource (QUrl): location of the input image
+        """
+        imgPath = imgSource.toLocalFile()
+        path = ThumbnailCache.thumbnailPath(imgPath)
+        logging.debug(f'[ThumbnailCache] Creating thumbnail {path} for image {imgPath}')
+
+        # Initialize image reader object
+        reader = QImageReader()
+        reader.setFileName(imgPath)
+        reader.setAutoTransform(True)
+
+        # Read image and check for potential errors
+        img = reader.read()
+        if img.isNull():
+            logging.error(f'[ThumbnailCache] Error when reading image: {reader.errorString()}')
+            return
+
+        # Scale image while preserving aspect ratio
+        thumbnail = img.scaled(ThumbnailCache.thumbnailSize, aspectMode=Qt.KeepAspectRatio)
+
+        # Write thumbnail to disk and check for potential errors
+        writer = QImageWriter(path)
+        success = writer.write(thumbnail)
+        if not success:
+            logging.error(f'[ThumbnailCache] Error when writing thumbnail: {writer.errorString()}')
+
+        # Notify listeners
+        self.thumbnailCreated.emit(imgSource)
+
+    @staticmethod
+    def clean():
+        """Scan the thumbnail directory and:
+        1. remove all thumbnails that have not been used for more than storageTimeLimit days
+        2. ensure that the number of thumbnails on disk does not exceed maxThumbnailsOnDisk.
+        """
+        # Check if thumbnail directory exists
+        if not os.path.exists(ThumbnailCache.thumbnailDir):
+            logging.debug('[ThumbnailCache] Thumbnail directory does not exist yet.')
+            return
+
+        # Get current time
+        now = time.time()
+
+        # Scan thumbnail directory and gather all thumbnails to remove
+        toRemove = []
+        remaining = []
+        for entry in os.scandir(ThumbnailCache.thumbnailDir):
+            if not entry.is_file():
```

Comment:
```suggestion
            if not entry.is_file(follow_symlinks=False):
```

Reply:
"The result is cached on the os.DirEntry object, with a separate cache for follow_symlinks True and False. Call [os.stat()](https://docs.python.org/3/library/os.html#os.stat) along with [stat.S_ISDIR()](https://docs.python.org/3/library/stat.html#stat.S_ISDIR) to fetch up-to-date information."

It will allow to have a single system call.